packages/core/saved-objects/core-saved-objects-api-server-internal/src/lib/repository.ts

```diff
@@ -762,6 +776,289 @@ export class SavedObjectsRepository implements ISavedObjectsRepository {
+    const { refresh = DEFAULT_REFRESH_SETTING, force } = options;
+    const namespace = normalizeNamespace(options.namespace);
+    const expectedBulkGetResults = this.presortObjectsByNamespaceType(objects);
+    const multiNamespaceDocsResponse = await this.preflightCheckForBulkDelete({
+      expectedBulkGetResults,
+      namespace,
+    });
+    const bulkDeleteParams: BulkDeleteParams[] = [];
+
+    const expectedBulkDeleteMultiNamespaceDocsResults =
+      this.getExpectedBulkDeleteMultiNamespaceDocsResults({
+        expectedBulkGetResults,
+        multiNamespaceDocsResponse,
+        namespace,
+        force,
+      });
+    // bulk up the bulkDeleteParams
+    expectedBulkDeleteMultiNamespaceDocsResults.map((expectedResult) => {
+      if (isRight(expectedResult)) {
+        bulkDeleteParams.push({
+          delete: {
+            _id: this._serializer.generateRawId(
+              namespace,
+              expectedResult.value.type,
+              expectedResult.value.id
+            ),
+            _index: this.getIndexForType(expectedResult.value.type),
+            ...getExpectedVersionProperties(undefined),
+          },
+        });
+      }
+    });
+
+    const bulkDeleteResponse = bulkDeleteParams.length
+      ? await this.client.bulk({
+          refresh,
+          body: bulkDeleteParams,
+          require_alias: true,
+        })
+      : undefined;
+
+    // extracted to ensure consistency in the error results returned
+    let errorResult: BulkDeleteItemErrorResult;
+    const objectsToDeleteAliasesFor: ObjectToDeleteAliasesFor[] = [];
+
+    const savedObjects = expectedBulkDeleteMultiNamespaceDocsResults.map((expectedResult) => {
+      if (isLeft(expectedResult)) {
+        return { ...expectedResult.value, success: false };
+      }
+      const {
+        type,
+        id,
+        namespaces,
+        esRequestIndex: esBulkDeleteRequestIndex,
+      } = expectedResult.value;
+      // we assume this wouldn't happen but is needed to ensure type consistency
+      if (bulkDeleteResponse === undefined) {
+        throw new Error(
+          `Unexpected error in bulkDelete saved objects: bulkDeleteResponse is undefined`
+        );
+      }
+      const rawResponse = Object.values(
+        bulkDeleteResponse.items[esBulkDeleteRequestIndex]
+      )[0] as NewBulkItemResponse;
+
+      const error = getBulkOperationError(type, id, rawResponse);
+      if (error) {
+        errorResult = { success: false, type, id, error };
+        return errorResult;
+      }
+      if (rawResponse.result === 'not_found') {
+        errorResult = {
+          success: false,
+          type,
+          id,
+          error: errorContent(SavedObjectsErrorHelpers.createGenericNotFoundError(type, id)),
+        };
+        return errorResult;
+      }
+
+      if (rawResponse.result === 'deleted') {
+        // `namespaces` should only exist in the expectedResult.value if the type is multi-namespace.
+        if (namespaces) {
```

Comment:
NIT: this condition should always be true here, right? I'm fine keeping, but then we should throw in a else block as we did for `if (bulkDeleteResponse === undefined) {`

Reply:
> this condition should always be true here, right?
Not necessarily. Only multi-namespace documents have a `namespaces` array because we're only using `mget` for those namespace types. Single namespace and agnostic types aren't fetched from es in the `mget` call.